jgrapht-ext/src/main/java/org/jgrapht/ext/GmlImporter.java

```diff
@@ -225,36 +229,62 @@ public void syntaxError(
         private Integer sourceId;
         private Integer targetId;
         private Double weight;
-
-        private Set<Integer> nodes;
-        private int singletons;
+        private String label;
+
+        /**
+         * Maps IDs, which are always integers, to the corresponding node labels
+         * for nodes that are connected in the graph.
+         */
+        private HashMap<Integer, String> nodeLabels;
+        /** Labels for stand-alone nodes that are not connected.  Note that these
+         * can be NULL.
+         */
+        private List<String> singletonLabels;
         private List<PartialEdge> edges;
 
         public void updateGraph(Graph<V, E> graph)
             throws ImportException
         {
             if (foundGraph) {
-                // add nodes
+                // Add connected nodes
                 int maxV = 1;
                 Map<Integer, V> map = new HashMap<Integer, V>();
-                for (Integer id : nodes) {
+                for (Integer id : nodeLabels.keySet()) {
+                	String label = nodeLabels.get(id);
+                	// Substitute the ID for the label if no label provided
+                	if (label == null || (label != null && label.equals(""))) {
+                		label = id + "";
+                	}
                     maxV = Math.max(maxV, id);
-                    V vertex =
-                        vertexProvider.buildVertex(id.toString(), new HashMap<String, String>());
-                    map.put(id, vertex);
-                    graph.addVertex(vertex);
+                    V vertex = vertexProvider.buildVertex(label, new HashMap<String, String>());
```

Comment:
Do not get confused by the naming of the attributes of the `VertexProvider` interface which follows
```
public interface VertexProvider<V>
{
    V buildVertex(String label, Map<String, String> attributes);
}
```
when you call the vertex provider, the first parameter must be the GML id and not the label. The newly added GML "label" is just an extra attribute and you must provide than to the vertex provider in the additional map called `attributes` (in case the label is not null of course).